### PR TITLE
Lambda delete confirm 219

### PR DESCRIPTION
**Pull request by micholas**
Added confirmation when deleting Lambda

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [x] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

* Added confirmation when deleting Lambda
* Removed use of globals from deleteLambda

## Motivation and Context

In response to issue.

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/219

## Testing

Updated test to:
* Check that delete is called when confirmed
* Check that delete is not called when cancelled
* Refresh is called on both confirm and cancel

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lambda/commands/deleteLambda.ts [hunk 1]
```diff
@@ -5,37 +5,63 @@
...
 import { LambdaClient } from '../../shared/clients/lambdaClient'
-import { ext } from '../../shared/extensionGlobals'
 import { StandaloneFunctionNode } from '../explorer/standaloneNodes'
 
-export async function deleteLambda(
-    node: StandaloneFunctionNode,
-    refresh: () => void
-): Promise<void> {
-    const client: LambdaClient = ext.toolkitClientBuilder.createLambdaClient(node.regionCode)
+export async function deleteLambda({
```
Comment (mpiroc) [inline src/lambda/commands/deleteLambda.ts]: Please fix the formatting/indentation:

```typescript
export async function deleteLambda({
    lambdaClient,
    ...
}: {
    lambdaClient: LambdaClient
    ....
}): Promise<void>
```

File: src/lambda/commands/deleteLambda.ts [hunk 2]
```diff
@@ -5,37 +5,63 @@
...
+                               return response === responseYes
+                           },
+                           onError = (error: any) => {
+                               outputChannel.show(true)
+                               outputChannel.appendLine(localize(
+                                   'AWS.command.deleteLambda.error',
+                                   "There was an error deleting lambda function '{0}'",
+                                   node.configuration.FunctionArn
+                                 ))
+                               outputChannel.appendLine(String(error)) // linter hates toString on type any
```
Comment (mpiroc) [inline src/lambda/commands/deleteLambda.ts]:
```typescript
onError = (error: Error) => {
    ...
}

...

catch (err) {
    onError(err as Error)
}
```

File: src/lambda/commands/deleteLambda.ts (hunk 1 above)
Comment (mpiroc) [inline src/lambda/commands/deleteLambda.ts]: When we chatted about destructured-objects-as-parameters, I didn't realize how much boilerplate it would add in TypeScript. Let's keep our codebase consistent for now--please switch to regular parameters.

File: src/lambda/commands/deleteLambda.ts [hunk 3]
```diff
@@ -5,37 +5,63 @@
...
-    node: StandaloneFunctionNode,
-    refresh: () => void
-): Promise<void> {
-    const client: LambdaClient = ext.toolkitClientBuilder.createLambdaClient(node.regionCode)
+export async function deleteLambda({
+                           lambdaClient,
+                           node,
+                           outputChannel,
+                           onRefresh,
+                           onConfirm = async () => {
```
Comment (mpiroc) [inline src/lambda/commands/deleteLambda.ts]: I don't think that this is the right level of mockability--we only want to mock `window.showWarningMessage`, the rest of `onConfirm` still needs test coverage.

File: src/lambda/commands/deleteLambda.ts [hunk 4]
```diff
@@ -5,37 +5,63 @@
...
+                                   "Are you sure you want to delete lambda function '{0}'?",
+                                   node.configuration.FunctionName
+                                 ),
+                                 responseYes,
+                                 responseNo
+                               )
+
+                               return response === responseYes
+                           },
+                           onError = (error: any) => {
```
Comment (mpiroc) [inline src/lambda/commands/deleteLambda.ts]: Similarly to `onConfirm`, we want to mock the output channel without losing test coverage of the rest of the error handler.

File: src/lambda/lambdaTreeDataProvider.ts [hunk 5]
```diff
@@ -42,21 +43,28 @@ export class LambdaTreeDataProvider implements vscode.TreeDataProvider<AWSTreeNo
...
         private readonly awsContextTrees: AwsContextTreeCollection,
         private readonly regionProvider: RegionProvider,
-        private readonly resourceFetcher: ResourceFetcher
+        private readonly resourceFetcher: ResourceFetcher,
+        private readonly  outputChannel: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Lambda')
     ) {
         this._onDidChangeTreeData = new vscode.EventEmitter<AWSTreeNodeBase | undefined>()
         this.onDidChangeTreeData = this._onDidChangeTreeData.event
         this.regionNodes = new Map<string, RegionNode>()
+        if (!ext.lambdaOutputChannel) {
```
Comment (mpiroc) [inline src/lambda/lambdaTreeDataProvider.ts]: Please don't do this piecemeal. It's also out-of-scope for this PR. In this PR, just use `ext.lambdaOutputChannel` instead of the constructor parameter. You can send a separate PR that removes `ext.lambdaOutputChannel` entirely and replaces it with constructor parameters.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 6]
```diff
@@ -43,20 +36,32 @@ describe('deleteLambda', async () => {
...
-        ext.toolkitClientBuilder = new MockToolkitClientBuilder(
-            undefined,
-            new MockLambdaClient(
-                undefined,
-                async name => {
-                    assert.strictEqual(name, 'myFunctionName')
-                    invokeCount++
-                }
-            )
+    const assertLambdaDeleteWorks = async ({confirmationResponse}: { confirmationResponse: boolean }) => {
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]:
```typescript
async function assertLambdaDeleteWorks(confirmationResponse: boolean) {
    ...
}
```

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 7]
```diff
@@ -76,29 +81,47 @@ describe('deleteLambda', async () => {
...
+        })
+
+        const expectedDeleteCount = confirmationResponse ? 1 : 0
+        assert.strictEqual(
+            deleteCount,
+            expectedDeleteCount,
+            `Expected delete count ${expectedDeleteCount}, actual count ${deleteCount}`
+        )
+
+        const expectedRefreshCount = 1
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Should we refresh even when the user declines the confirmation?

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 8]
```diff
@@ -76,29 +81,47 @@ describe('deleteLambda', async () => {
...
-        await deleteLambda(node, () => {})
+        await deleteLambda({
+             node,
+             lambdaClient,
+             outputChannel: new MockOutputChannel(),
+             onRefresh: () => refreshCount++,
+             onConfirm: async () => confirmationResponse
+        })
+
+        const expectedDeleteCount = confirmationResponse ? 1 : 0
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: It's easier to read/maintain the tests if the tests themselves set the expectations. Please pass in `expected*Count` from the individual tests.

File: src/test/mockOutputChannel.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  public hide(): void {
+    this.isHidden = true
+  }
+
+  /**
+   * This is overloaded by VS Code but viewColumn version is deprecated thus
+   * show(preserveFocus?: boolean) is really what we should consider.
+   * @param args
+   */
+  public show(...args: any[] /* viewColumn?: ViewColumn, preserveFocus?: boolean */) {
```
Comment (mpiroc) [inline src/test/mockOutputChannel.ts]: See the 'Overloads' section [here](https://www.typescriptlang.org/docs/handbook/functions.html) for the correct way to implement overloads in TypeScript.

File: src/lambda/commands/deleteLambda.ts [hunk 10]
```diff
@@ -5,37 +5,63 @@
 
 'use strict'
 
-import * as nls from 'vscode-nls'
-const localize = nls.loadMessageBundle()
+import { OutputChannel, window } from 'vscode'
+import { loadMessageBundle } from 'vscode-nls'
```
[the remark this answers is not in the document]
Comment (micholas) [reply src/lambda/commands/deleteLambda.ts]: I'd like to start working towards eventually using WebPack to make our extension smaller. In my experience with Lambda, using WebPack enabled us to reduce the size by a factor of 100 to 1000 from the MB range to the KB range. Webpack treeshaking leverages ECMA6 style imports to prune code that's not used.

[the remark this answers is not in the document]
Comment (mpiroc) [reply src/lambda/commands/deleteLambda.ts]: Treeshaking is smart enough to handle `import *`. See https://github.com/webpack/webpack/tree/master/examples/harmony-unused#examplejs.

[the remark this answers is not in the document]
Comment (micholas) [reply src/lambda/commands/deleteLambda.ts]: OK. I'll follow the convention.

File: src/lambda/commands/deleteLambda.ts (hunk 1 above)
Comment (micholas) [reply src/lambda/commands/deleteLambda.ts]: Fixed

Comment (micholas) [reply src/lambda/commands/deleteLambda.ts]: Removed "boiler plate" of repeated parameter names in favor of a single restParams parameter. Long lists of arguments to a function/constructor has been considered poor practice in many languages for a long time. This is just as true in TypeScript as it is in JavaScript and Java.

https://medium.com/@janakachathuranga/coding-best-practices-javascript-9a5f397d701b

Look at this example: 
```
                new AWSCommandTreeNode(
                    undefined,
                    localize('AWS.explorerNode.signIn', 'Connect to AWS...'),
                    'aws.login',
                    undefined,
                    localize('AWS.explorerNode.signIn.tooltip', 'Connect to AWS using a credential profile'))
```
What do each of those arguments represent? You may know but the typical reader does not.

By passing a single typed object, the intent is obvious:
```
            async (node: StandaloneFunctionNode) => await deleteLambda({
                node,
                lambdaClient: ext.toolkitClientBuilder.createLambdaClient(node.regionCode),
                outputChannel: ext.lambdaOutputChannel,
                onRefresh: () => this.refresh(node.parent)
})
```
I also feel this CR is a reasonable place for this change. If a function has one argument, that's fine, two maybe, but when a CR add a 3rd or 4th it should change the function to take a single typed object as an argument.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 8 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: No `assertLambdaDeleteWorks` is a function which encapsulates logic to verify that that deleting works when expected. Expected delete count, refresh count, and any applicable output verified are internal implementation detail of the helper function. Prefer not to have leaky abstraction.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 7 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: That is a good question. I haven't participated in a discussion about when we want to refresh the list of Lambdas but I'd be fine either way.

File: src/lambda/commands/deleteLambda.ts (hunk 3 above)
Comment (micholas) [reply src/lambda/commands/deleteLambda.ts]: Our need to mock can be greatly reduced by writing less brittle, tightly coupled code. Thinking in a non-tightly coupled way, we are inserting logic to support confirmation. It might be sending and SMS, receiving and verifying the associated security token, it may be a capcha... the implementation really doesn't matter. What does matter is that we are adding support for a function that enables confirm or deny by simply returning true of false.

See "Mocking as a code smell": https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a

Yes we could test `vscode.window.showWarningMessage` as part of our integ tests, but the need to mock it only stems from tightly coupled code.

File: src/test/mockOutputChannel.ts (hunk 9 above)
Comment (micholas) [reply src/test/mockOutputChannel.ts]: Fixed the types by defining them in a manner similar to VS Code.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 7 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: It's probably better not to refresh when the user doesn't confirm since it might cause the Lambdas displayed to shift which might confuse the user who is not expecting a change. Fixed to behave this way.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 8 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Took your point to heart. Ultimately it would be ideal if we could quickly see the input and output factors of the test cases at a glance without having to dig through the code. Re-factored to do so.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 11]
```diff
@@ -6,158 +6,160 @@
...
+        }),
+        ...params
+    }: {
+        functionName: string,
+        errorToThrowDuringDelete?: Error,
+        expectDeleteIsCalled: boolean,
+        expectRefreshIsCalled: boolean,
+        onConfirm(): Promise<boolean>,
+        onAssertOutputChannel?(actualOutputChannel: MockOutputChannel): void
+    }) => {
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Throughout this file, indentation should be 4 spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 12]
```diff
@@ -6,158 +6,160 @@
...
+              channel.value,
+              '',
+              'expect no output since output testing was omitted'
             )
+        }),
+        ...params
+    }: {
+        functionName: string,
+        errorToThrowDuringDelete?: Error,
+        expectDeleteIsCalled: boolean,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Nit: I think it's better to use counts rather than booleans, like in your previous iteration. That way, we're guarding against redundant calls.

(the correct solution is to use a real mocking framework, but in the meantime...)

File: src/test/mockOutputChannel.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+  public hide(): void {
+    this.isHidden = true
+  }
+
+  /**
+   * This is overloaded by VS Code but viewColumn version is deprecated thus
+   * show(preserveFocus?: boolean) is really what we should consider.
+   */
+  public show(columnOrPreserveFocus?: vscode.ViewColumn | boolean, preserveFocus?: boolean): void {
```
Comment (mpiroc) [inline src/test/mockOutputChannel.ts]: Should we throw on `preserveFocus !== undefined`?

Comment (mpiroc) [inline src/test/mockOutputChannel.ts]: The implementation is good--for consistency, please add the overload signatures above it:

```typescript
public show(preserveFocus?: boolean): void
public show(column?: ViewColumn, preserveFocus?: boolean): void
public show(columnOrPreserveFocus?: vscode.ViewColumn | boolean, preserveFocus?: boolean): void {
    // ...
}
```

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 12 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Good point, didn't think about that. Conceptually `expectDeleteIsCalledOnce: boolean`

File: src/test/mockOutputChannel.ts (hunk 13 above)
Comment (micholas) [reply src/test/mockOutputChannel.ts]: I don't think so. It is optional. My understanding is that regardless of the the arguments, calling `show` will un-hide the output channel and that preserving focus is tangential which is why all params are optional in the `OutputChannel` interface.

Comment (micholas) [reply src/test/mockOutputChannel.ts]: That exists in the OutputChannel interface defined by VS Code, which `MockOutputChannel` formally implements.

https://code.visualstudio.com/api/references/vscode-api#OutputChannel
https://github.com/Microsoft/vscode/blob/7c1fe9a7d6eaa6b3bbfddb740996c4bb867e6385/src/vs/vscode.d.ts

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 11 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Thanks. It's been a constant battle against my IDE. Need to fix my defaults. Will address the code in a follow on CR.

Comment (mpiroc) [reply src/test/lambda/commands/deleteLambda.test.ts]: Thanks--please address in this PR rather than a follow-up. Better to avoid checking in misformatted code than to go back and fix it.

Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Done.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 14]
```diff
@@ -6,158 +6,160 @@
...
+            onConfirm: async () => true,
+
+            // expected results
+            expectDeleteIsCalled: true,
+            expectRefreshIsCalled: true,
+            onAssertOutputChannel(outputChannel: MockOutputChannel) {
+                const expectedMessagePart = String(errorToThrowDuringDelete)
+                assert(!outputChannel.isHidden, 'output channel should not be hidden after error')
+                assert(
+                  outputChannel.value && outputChannel.value.indexOf(expectedMessagePart) > 0,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 15]
```diff
@@ -6,158 +6,160 @@
...
-                undefined,
-                async name => {
-                    deleteCount++
-                    throw new Error()
-                }
+    const assertLambdaDeleteWorksWhen = async ({
+        onAssertOutputChannel = ((channel: MockOutputChannel) => {
+            // Defaults to expecting no output. Should verify output when expected.
+            assert.strictEqual(
+              channel.value,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 16]
```diff
@@ -6,158 +6,160 @@
...
+        errorToThrowDuringDelete?: Error,
+        expectDeleteIsCalled: boolean,
+        expectRefreshIsCalled: boolean,
+        onConfirm(): Promise<boolean>,
+        onAssertOutputChannel?(actualOutputChannel: MockOutputChannel): void
+    }) => {
+        let isDeleteCalled = false
+        let isRefreshCalled = false
+        const lambdaClient = new MockLambdaClient(
+          undefined,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 17]
```diff
@@ -6,158 +6,160 @@
...
+          async (name) => {
+              isDeleteCalled = true
+              assert.strictEqual(
+                name,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 18]
```diff
@@ -6,158 +6,160 @@
...
-            show(preserveFocus?: boolean | undefined): void { },
-            appendLine(value: string): void { }
-        } as any as OutputChannel
-
         const parent = new MockStandaloneFunctionGroupNode(
-            undefined,
-            undefined,
-            async () => assert.fail(),
-            async () => assert.fail()
+          undefined,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 19]
```diff
@@ -6,158 +6,160 @@
...
         const node = new MockStandaloneFunctionNode(
-            undefined,
-            parent,
-            {
-                FunctionArn: 'myFunctionArn',
-                FunctionName: 'myFunctionName'
-            },
-            async () => assert.fail(),
-            async configuration => assert.fail()
+          undefined,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 20]
```diff
@@ -6,158 +6,160 @@
...
+            if (params.errorToThrowDuringDelete) {
+                assert.deepStrictEqual(params.errorToThrowDuringDelete, err)
+            } else {
+                console.error(error)
+                assert.fail(`Unexpected error during test: "${error.message}"`)
+            }
+        }
+
+        assert.strictEqual(
+          isDeleteCalled,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/lambda/commands/deleteLambda.test.ts [hunk 21]
```diff
@@ -6,158 +6,160 @@
...
-            undefined,
-            parent,
-            {
-                FunctionArn: 'myFunctionArn',
-                FunctionName: 'myFunctionName'
-            },
-            async () => assert.fail(),
-            async configuration => assert.fail()
+        assert.strictEqual(
+          isRefreshCalled,
```
Comment (mpiroc) [inline src/test/lambda/commands/deleteLambda.test.ts]: Four spaces.

File: src/test/mockOutputChannel.ts [hunk 22]
```diff
@@ -0,0 +1,47 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+export class MockOutputChannel implements vscode.OutputChannel {
+  public value: string = ''
```
Comment (mpiroc) [inline src/test/mockOutputChannel.ts]: Four spaces.

File: src/test/mockOutputChannel.ts [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  public isHidden: boolean = false
+  public preserveFocus: boolean = false
+
+  public readonly name = 'Mock channel'
+
+  public append(value: string): void {
+    this.value += value
```
Comment (mpiroc) [inline src/test/mockOutputChannel.ts]: Four spaces.

Comment (mpiroc) [reply src/test/mockOutputChannel.ts]: Same throughout this file.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 11 above)
Comment (mpiroc) [reply src/test/lambda/commands/deleteLambda.test.ts]: There are many other places where indentation needs to be fixed. I added comments to each that I saw.

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 12 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Will do next CR

File: src/test/mockOutputChannel.ts (hunk 23 above)
Comment (micholas) [reply src/test/mockOutputChannel.ts]: Will do next CR

File: src/test/mockOutputChannel.ts (hunk 22 above)
Comment (micholas) [reply src/test/mockOutputChannel.ts]: Will do next CR

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 21 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Will do next CR

File: src/test/lambda/commands/deleteLambda.test.ts (hunk 20 above)
Comment (micholas) [reply src/test/lambda/commands/deleteLambda.test.ts]: Will do next CR
